boards: arm: nrf9160_pca10090: choose a code-partition and fix offset

lemrey opened the pull request:
- Choose a `code-partition` for the secure board variant
- Introduce the BOARD_NRF9160_PCA10090NS
- Set `FLASH_LOAD_OFFSET` and `FLASH_LOAD_SIZE` in the non-secure board variant defconfig file

This patch fixes building/flashing both secure and nonsecure firmware images, broken after:  9ce4cce3493c53f01e2e5f454a36a3149cd153a5
Please @ioannisg, have a look.

ioannisg reviewed (commented):
Blocking this until we know what to do with TZ-enabled boards.

lemrey commented:
@ioannisg , updated this as discussed.

galak commented on boards/arm/nrf9160_pca10090/nrf9160_pca10090ns_defconfig:
```diff
@@ -12,6 +12,8 @@ CONFIG_ARM_TRUSTZONE_M=y
 
 # This Board implies building Non-Secure firmware
 CONFIG_TRUSTED_EXECUTION_NONSECURE=y
+CONFIG_FLASH_LOAD_OFFSET=0x40000
```
If this value is based on slot0_partition, than why not get it from DT like we do MCUBOOT?

ioannisg replied:
Agree with Kumar, we should do this. It should be based on slot0_nonsecure, if I am not wrong.

ioannisg replied:
Can't we take DT_CODE_PARTITION_OFFSET / SIZE ?

lemrey replied:
I had tried: `CONFIG_FLASH_LOAD_OFFSET = DT_CODE_PARTITION_OFFSET`, didn't work.

galak replied:
You can do what you did initially, but just put it in the board Kconfig.defconfig for now.  (Take a look at boards/arm/mimxrt1050_evk/Kconfig.defconfig for example).

ioannisg replied:
Would it work if you use: $(dt_hex_val,DT_CODE_PARTITION_OFFSET)?

galak replied:
Yes that should work

ioannisg replied:
Otherwise, sure, we can always add `select FLASH_LOAD_OFFSET=$(dt_hex_val,DT_CODE_PARTITION_OFFSET)` inside Kconfig.defconfig (assuming that _this_ would work) :)

lemrey replied:
> Otherwise, sure, we can always add `select FLASH_LOAD_OFFSET=$(dt_hex_val,DT_CODE_PARTITION_OFFSET)` inside Kconfig.defconfig (assuming that _this_ would work) :)

I doubt `select` work, you can only `select` a boolean entry, afaik. 
Kumar suggestion does work, but the board Kconfig.defconfig is shared between secure and non-secure variants, so we'd need something like:

`config FLASH_LOAD_OFFSET`
`	default $(dt_hex_val,DT_CODE_PARTITION_OFFSET) if CONFIG_TRUSTED_EXECUTION_NONSECURE`

ioannisg replied:
The "if" part is acceptable, I think.

 So, `
CONFIG_FLASH_LOAD_OFFSET=FLASH_LOAD_OFFSET=$(dt_hex_val,DT_CODE_PARTITION_OFFSET)` in the defconfig of the non-secure does not work, @lemrey?

lemrey replied:
> So, ` CONFIG_FLASH_LOAD_OFFSET=FLASH_LOAD_OFFSET=$(dt_hex_val,DT_CODE_PARTITION_OFFSET)` in the defconfig of the non-secure does not work, @lemrey?

No:
`CONFIG_FLASH_LOAD_OFFSET=$(dt_hex_val,DT_CODE_PARTITION_OFFSET)`
cannot be placed in `boards/arm/nrf9160_pca10090/nrf9160_pca10090ns_defconfig`

It needs to go in `boards/arm/nrf9160_pca10090/Kconfig.defconfig`

lemrey replied:
@galak Updated as discussed.

lemrey committed:
boards: arm: nrf9160_pca10090: choose code-partition in secure variant

Choose a code partition in the secure board variant
for consistency with other boards.

Signed-off-by: Emanuele Di Santo <emdi@nordicsemi.no>

ioannisg commented on boards/arm/nrf9160_pca10090/nrf9160_pca10090.dts:
```diff
@@ -12,5 +12,6 @@
 	chosen {
 		zephyr,sram = &sram0;
 		zephyr,flash = &flash0;
+		zephyr,code-partition = &slot0_partition;
```
This is fine. Aligns with the rest of the ARM boards.

ioannisg commented on boards/arm/nrf9160_pca10090/Kconfig.defconfig:
```diff
@@ -30,4 +30,10 @@ config SPI_3
 
 endif # SPI
 
+config FLASH_LOAD_OFFSET
```
If I have understood correctly how Kconfig works, this definition overwrites the default definition in Kconfig.zephyr, right? Do not we need to have the same content here, too? I.e. the content below (copy-pasted)
```
hex "Kernel load offset"
default 0
depends on HAS_FLASH_LOAD_OFFSET
```

ioannisg replied:
And same for _SIZE, of course

lemrey replied:
Apparently not, we don't need to repeat any of that; we can just override the default.
I checked in `menuconfig` and everything looks okay, the help message is there too.
Perhaps @ulfalizer can comment?

ulfalizer replied:
Defining a symbol/choice multiple times merges the definitions. You get a symbol with all the properties from the different locations, and the dependencies get ORed.

Take this definition for example:

```
config FOO
	int "foo"
	default 1 if A
	default 2 if B
	depends on X

...

config FOO
	default 3 if C
	depends on Y

...

config FOO
	default 4 if D
	default 5
	depends on Z
```

It's logically equivalent to this:

```
config FOO
	int "foo"
	default 1 if A
	default 2 if B
	default 3 if C
	default 4 if D
	default 5
	depends on X || Y || Z
```

There are some other subtleties as well. For example, if you put a prompt on more than one of the definitions, then you'll be able to see the symbol in multiple places in the `menuconfig` (wherever the prompt is visible). Internally, it's still a single symbol though.

lemrey replied:
@ioannisg I think we can leave this as-is then.

ioannisg replied:
ok

ioannisg commented on boards/arm/nrf9160_pca10090/Kconfig.defconfig:
```diff
@@ -30,4 +30,10 @@ config SPI_3
+config FLASH_LOAD_OFFSET
+	default $(dt_hex_val,DT_CODE_PARTITION_OFFSET) if TRUSTED_EXECUTION_NONSECURE
+
+config FLASH_LOAD_SIZE
```
It would be nice to explain with a comment, why this Kconfigs are required for the Non-Secure build.

For example we need to clarify that we do not have any more DTS overlays (as we had for MCUboot) and so the linker gets the OFFSET and SIZE from Kconfig (which needs _this kind of_ setup, to obtain it from DTS)

ioannisg replied:
And same for _OFFSET of course.

lemrey replied:
Updated

ioannisg replied:
Text is nice.

ioannisg reviewed (changes requested):
Looks good and it works - I added few comments.

However, I realized that it would be better, if the "if" conditional contains the BOARD Kconfig, and not TRUSTED_EXECUTION_NONSECURE.
Because, it is the _specific board_ that requires this offsetting; it is **not** the fact that this is a Non-Secure Firmware. It will be more readable, IMO.

So, I would put `if CONFIG_BOARD_NRF9160_PCA10090NS`.
Now, I realize that we currently have the same `CONFIG_BOARD_` symbol for both the default and the non-secure boards. So, to apply my suggestion, we need to provide the infastructure for the 2 board Kconfig symbols.

@galak agree? @lemrey  could you do it?

lemrey committed:
boards: arm: nrf9160_pca10090: define BOARD_NRF9160_PCA10090NS

Introduce BOARD_NRF9160_PCA10090NS to differentiate between
secure and non-secure board variants.

Signed-off-by: Emanuele Di Santo <emdi@nordicsemi.no>

lemrey committed:
boards: arm: nrf9160_pca10090: link into code-partition if non-secure

Always link firmware built for nrf9160_pca10090ns into the
code-partition defined in DT, regardless.

Signed-off-by: Emanuele Di Santo <emdi@nordicsemi.no>

lemrey commented:
@ioannisg I have addressed your comments and introduced `BOARD_NRF9160_PCA10090NS`

ioannisg commented on boards/arm/nrf9160_pca10090/Kconfig.board:
```diff
@@ -4,6 +4,14 @@
 config BOARD_NRF9160_PCA10090
 	bool "nRF9160 PCA10090"
-	depends on SOC_NRF9160_SICA
+	depends on !TRUSTED_EXECUTION_NONSECURE
```
You could skip this dependence (although it is correct). It is not useful, I think. The default value will be false, anyways.

lemrey replied:
My intention was to keep only one choice available. If that condition is omitted, then in case of non-secure firmware both the secure and non-secure board versions are shown in menuconfig. Although the right one will be selected, I'd rather avoid that.

ioannisg replied:
Could there be a choice?

lemrey replied:
No, there is no choice; when building for secure the board is nrf9160_pca10090, when building for non-secure the board is nrf9160_pca10090ns. So removing the dependency will let show a choice that is wrong.

ioannisg commented on boards/arm/nrf9160_pca10090/Kconfig.board:
```diff
@@ -4,6 +4,14 @@
+	bool "nRF9160 PCA10090 non-secure"
+	depends on TRUSTED_EXECUTION_NONSECURE
+
+endif # SOC_NRF9160_SICA
```
Something with the end-of-line?

lemrey replied:
Fixed

ioannisg commented on boards/arm/nrf9160_pca10090/Kconfig:
```diff
@@ -1,9 +0,0 @@
-# Kconfig - nRF9160 PCA10090 board configuration
```
I would prefer not to delete this file even if it is empty. AFAIK we have similar "empy" files in other board definitions. You could add an empty if..endif for the _NS board too, to make it consistent :)

lemrey replied:
Kept it and updated it.

ioannisg reviewed (approved):
LGTM
